packages/annotations/src/vote.ts

```diff
@@ -0,0 +1,125 @@
+/* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
+ * Apache-2.0 */
+
+import {
+  ArcGISRequestError,
+  ArcGISAuthError,
+  IRequestOptions
+} from "@esri/arcgis-rest-request";
+
+import {
+  addFeatures,
+  deleteFeatures,
+  queryFeatures,
+  IAddFeaturesRequestOptions,
+  IDeleteFeaturesRequestOptions,
+  IAddFeaturesResult,
+  IDeleteFeaturesResult,
+  IQueryFeaturesResponse
+} from "@esri/arcgis-rest-feature-service";
+
+import { IAnnoFeature } from "./add";
+import { UserSession } from "@esri/arcgis-rest-auth";
+
+export interface IVoteRequestOptions extends IRequestOptions {
+  url: string;
+  annotation: IAnnoFeature;
+  /**
+   * Set to 'true' in order to vote another annotation down.
+   */
+  downVote?: boolean;
+  authentication: UserSession;
+}
+
+/**
+ * ```js
+ * import { vote } from "@esri/hub-annotations";
+ * //
+ * voteOnAnnotation({
+ *   url,
+ *   annotation: { attributes: {} },
+ *   authentication
+ * })
+ *   .then(response);
+ * ```
+ * Upvote (or downvote) another annotation.
+ * @param requestOptions - request options that may include authentication
+ * @returns A Promise that will resolve with the response from the service after attempting to vote on an annotation.
+ */
+export function voteOnAnnotation(
+  requestOptions: IVoteRequestOptions
+): Promise<IAddFeaturesResult | IDeleteFeaturesResult> {
+  const annotation = requestOptions.annotation;
+  const auth = requestOptions.authentication as UserSession;
+  const url = requestOptions.url;
+
+  if (!auth) {
+    return new Promise((resolve, reject) => {
+      reject(
+        new ArcGISAuthError(`Voting by anonymous users is not supported.`)
+      );
+    });
+  }
+
+  if (annotation.attributes.author === auth.username) {
+    return new Promise((resolve, reject) => {
+      reject(
+        new ArcGISRequestError(`Users may not vote on their own comment/idea.`)
+      );
+    });
+  }
+
+  // searchAnnotations() would make more xhrs for user metadata
+  return queryFeatures({
+    url,
+    where: `parent_id = ${annotation.attributes.OBJECTID} AND author = '${
+      auth.username
+    }'`
+  }).then((
+    queryResponse: any /* why can't i declare it as IQueryFeaturesResponse ? */
```

Comment:
here's another one headscratcher.

`queryFeatures()` returns a union type, yet i can't explicitly cast a response i know will be `IQueryFeaturesResponse` (and not `IQueryResponse`) to introspect properties that i **know** will be available, so i've resorted to yet another `any`.

my TS foo is too weak.

Reply:
by cast, did you mean `(queryResponse as IQueryFeaturesResponse)`?

Reply:
yup.  	

> Argument of type 'IQueryFeaturesResponse' is not assignable to parameter of type '(value: IQueryFeaturesResponse | IQueryResponse) => IQueryFeaturesResponse | IQueryResponse | PromiseLike<IQueryFeaturesResponse | IQueryResponse>'.
  Type 'IQueryFeaturesResponse' provides no match for the signature '(value: IQueryFeaturesResponse | IQueryResponse): IQueryFeaturesResponse | IQueryResponse | PromiseLike<IQueryFeaturesResponse | IQueryResponse>

🤔